### PR TITLE
`ClusterSettings` does not compute `ClusterConnectionMode` consistently

### PR DESCRIPTION
**Pull request by stIncMale**
This PR addresses test concerns expressed in the original PR https://github.com/mongodb/mongo-java-driver/pull/1163. New tests discovered more issues, so more changes were done in `ClusterSettings`.

JAVA-5088

### CODE REVIEW

File: driver-core/src/main/com/mongodb/connection/ClusterSettings.java [hunk 1]
```diff
@@ -337,16 +337,15 @@ public Builder applyConnectionString(final ConnectionString connectionString) {
...
+                mode(directConnection ? ClusterConnectionMode.SINGLE : ClusterConnectionMode.MULTIPLE);
+                hosts(singletonList(createServerAddress(connectionString.getHosts().get(0))));
             } else {
+                mode = null;
                 List<ServerAddress> seedList = connectionString.getHosts().stream()
                         .map(ServerAddressHelper::createServerAddress)
                         .collect(Collectors.toList());
-                mode(ClusterConnectionMode.MULTIPLE).hosts(seedList);
+                hosts(seedList);
             }
```
Comment (stIncMale) [inline driver-core/src/main/com/mongodb/connection/ClusterSettings.java:349]: Without this change, the following test fails:

```java
        when:
        settings = ClusterSettings.builder()
                .applyConnectionString(new ConnectionString('mongodb://example.com:27018'))
                .requiredReplicaSetName("test")
                .build()

        then:
        settings.mode == ClusterConnectionMode.MULTIPLE
```

File: driver-core/src/main/com/mongodb/connection/ClusterSettings.java [hunk 2]
```diff
@@ -337,16 +337,15 @@ public Builder applyConnectionString(final ConnectionString connectionString) {
...
                     srvServiceName(srvServiceName);
                 }
-            } else if ((directConnection != null && directConnection)
-                    || (directConnection == null && connectionString.getHosts().size() == 1
-                        && connectionString.getRequiredReplicaSetName() == null)) {
-                mode(ClusterConnectionMode.SINGLE)
-                        .hosts(singletonList(createServerAddress(connectionString.getHosts().get(0))));
+            } else if (directConnection != null) {
+                mode(directConnection ? ClusterConnectionMode.SINGLE : ClusterConnectionMode.MULTIPLE);
+                hosts(singletonList(createServerAddress(connectionString.getHosts().get(0))));
```
[the remark this answers is not in the document]
Comment (jyemin) [reply driver-core/src/main/com/mongodb/connection/ClusterSettings.java:342]: Confirmed that this test fails on the branch and passes on main branch:

```
        when:
        settings = ClusterSettings.builder()
                .applyConnectionString(new ConnectionString('mongodb://example.com:27017,example.com:27018/?directConnection=false'))
                .build()

        then:
        settings.mode == ClusterConnectionMode.MULTIPLE
        settings.hosts == [new ServerAddress('example.com:27017'), new ServerAddress('example.com:27018')]
```

[the remark this answers is not in the document]
Comment (stIncMale) [reply driver-core/src/main/com/mongodb/connection/ClusterSettings.java]: Thanks! Yes, we should not ignore the rest of the seed hosts if `directConnection` is `false`.

File: driver-core/src/test/unit/com/mongodb/connection/ClusterSettingsSpecification.groovy [hunk 3]
```diff
@@ -192,6 +205,20 @@ class ClusterSettingsSpecification extends Specification {
...
         settings.srvServiceName == 'mongodb'
 
+        when:
+        settings = ClusterSettings.builder()
+                .applyConnectionString(new ConnectionString('mongodb://example.com:27018'))
+                .requiredReplicaSetName("test")
+                .build()
+
+        then:
+        settings.mode == ClusterConnectionMode.MULTIPLE
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/connection/ClusterSettingsSpecification.groovy:215]: A note for us in the future: when run against the original code, this assertion fails

File: driver-core/src/test/unit/com/mongodb/connection/ClusterSettingsSpecification.groovy [hunk 4]
```diff
@@ -288,15 +330,16 @@ class ClusterSettingsSpecification extends Specification {
...
     }
 
-    def 'when cluster type is unknown and replica set name is specified, should set cluster type to ReplicaSet'() {
+    def 'when cluster type is UNKNOWN and replica set name is set, should set cluster type to REPLICA_SET and mode to MULTIPLE'() {
         when:
         def settings = ClusterSettings.builder().hosts([new ServerAddress()]).requiredReplicaSetName('yeah').build()
 
         then:
         ClusterType.REPLICA_SET == settings.requiredClusterType
+        ClusterConnectionMode.MULTIPLE == settings.mode
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/connection/ClusterSettingsSpecification.groovy:339]: A note for us in the future: when run against the original code, this assertion fails

**Review by jyemin — APPROVED** [review]
Even in relatively simple code like this, I don't find it easy to tease apart the refactoring from the behavioral change that fixes the bug, and would probably prefer that it be split into two commits.  But I think I understand everything and the tests look good, so LGTM regardless.

**Review by vbabanin — APPROVED** [review]

**Pull request closed without merging**